PXC-702: GCache encryption

kamil-holubicki opened the pull request:
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Server part implementation.
1. Needed changes around keyring component and Galera initialization
2. Master key rotation

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Server part implementation.
1. Needed changes around keyring component and Galera initialization
2. Master key rotation

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Galera repo pointer updated

dutow commented on plugin/keyring/common/keys_container.cc:
```diff
@@ -183,8 +206,11 @@ bool Keys_container::load_keys_from_keyring_storage() {
-        was_error = true;
-        delete key_loaded;
+#ifdef WITH_WSREP
+        if (forceKeysFetch) {
```
Shouldn't `else` keep existing behavior?

kamil-holubicki replied:
The part related to this hack has been completely reworked. Now Keyring plugin is reloaded after SST.

dutow commented on plugin/keyring/common/keys_container.cc:
```diff
@@ -88,6 +92,12 @@ bool Keys_container::store_key_in_hash(IKey *key) {
 bool Keys_container::store_key(IKey *key) {
+#ifdef WITH_WSREP
+  if (forceKeysFetch) {
+    load_keys_from_keyring_storage();
```
These hacks ignore the return value of load_keys_from_keyring_storage, which might fail, especially with a remote keyring.

kamil-holubicki replied:
The part related to this hack has been completely reworked. Now Keyring plugin is reloaded after SST.

dutow reviewed (commented):
This seems too much of a hack to me.

Can't we explicitly reload the keyring cache after SST instead?

kamil-holubicki committed:
PXC-4014: During crash recovery server is unable
to find keyring to decrypt tablespaces

Fixed the issue.

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Galera repo pointer updated.

Galera unit tests:
1. compilation fixed
2. adjusted UTs to GCache encryption changes

kamil-holubicki committed:
PXC-4014: During crash recovery server is unable
to find keyring to decrypt tablespaces

Got rid of extern lobal hack for kering reload.
Plugin reinitialization introduced.

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Integrated existing MTR test suites in the way that they can be executed
with encryption enabled or disabled. Used combinations file for this and
two auxiliary variables:
wsrep_gcache_encrypt
wsrep_disk_pages_encrypt

Using of these variables is necessary, because wserp_provider_params
cannot be extended via combinations file.

wsrep_deinit() call added to unireg_abort() (MK infrastructure
deinitialization)

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

auxiliary variables:
wsrep_gcache_encrypt
wsrep_disk_pages_encrypt
are read only now.

Improved galera_defaults test.

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Allow SST script to work. It needs secure connection when keyring plugin
is loaded.

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Galera repo pointer updated

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

MTR tests modifications

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Galera submodule pointer updated.

kamil-holubicki committed:
PXC-4027: GCache Master Key Rotation Fails on some instances

https://jira.percona.com/browse/PXC-4027

1. Master Key generation failure logging moved to the client side as its
logic has been changed.

2. Galera repo pointer updated

kamil-holubicki committed:
PXC-4031: Rotation of GCache Master Key crashes the server

https://jira.percona.com/browse/PXC-4031

Galera repo pointer updated

kamil-holubicki committed:
PXC-4035: Add a message in server logs when GCache Master key is rotated

https://jira.percona.com/browse/PXC-4035

Galera repo pointer updated.

kamil-holubicki committed:
PXC-4039: GCache MK is not replicated from source to replica nodes

https://jira.percona.com/browse/PXC-4039

Added replication of ALTER INSTANCE ROTATE GCACHE MASTER KEY

kamil-holubicki committed:
PXC-4036: Server killed during GCache MK rotation causes failures after server restart.

https://jira.percona.com/browse/PXC-4036

Updated Galera repo pointer

kamil-holubicki committed:
PXC-4044: Replicating GCache MK breaks the cluster if the node has gcache encryption disabled

https://jira.percona.com/browse/PXC-4044

Galera repo pointer updated.

kamil-holubicki committed:
wsrep-lib, wsrep-api, galera parts rebased on top
of heads (galera 26.4.12). Server part still waits
to be rebased on top of 8.0.30.

kamil-holubicki committed:
PXC-4040: Store GCache Master Key in database table

https://jira.percona.com/browse/PXC-4040

Extended string length in keyring_keys to the full capacity of the field

kamil-holubicki committed:
galera repo pointer updated

kamil-holubicki committed:
galera repo pointer updated
wsrep-lib repo pointer updated
